custom/aaa/models.py

```diff
@@ -0,0 +1,272 @@
+"""Contains all models for the AAA Convergence Dashboard.
+
+Conventions followed in this file:
+    * Fields ending in _ranges represent a field's history from forms that are useful for filtering.
+      This is only used for fields that are required to filter for specific ranges (usually a date range).
+    * Fields ending in _history represent a field's entire history from forms.
+      They are saved with the format [[date, value], [date, value]].
+      They are not guaranteed to be ordered by date.
+      They are not intended to be queried on, their interpretation should happen in python.
+      These history fields are saved in a separate model to ensure that the main model stays
+      fast and small for filtering.
+    * Each model contains all parts of the location hierarchy.
+    * Each model should support a future world where it is being accessed in a
+      sharded database (CitusDB).
+"""
+
+from __future__ import absolute_import, unicode_literals
+
+from django.contrib.postgres.fields import ArrayField, DateRangeField
+from django.db import models
+
+from corehq.apps.userreports.models import StaticDataSourceConfiguration, get_datasource_config
+from corehq.apps.userreports.util import get_table_name
+
+
+class LocationDenormalizedModel(models.Model):
+    """Abstract base model for containing all the location fields necessary for querying."""
+
+    # support multiple domains so QA is easier
+    domain = models.TextField(null=False)
+
+    # Shared location hierarchy
+    state_id = models.TextField(null=True)
+    district_id = models.TextField(null=True)
+
+    # CAS Hierarchy
+    block_id = models.TextField(null=True)
+    supervisor_id = models.TextField(null=True)
+    awc_id = models.TextField(null=True)
+
+    # AAA/MoH Hierarchy
+    taluka_id = models.TextField(null=True)
+    phc_id = models.TextField(null=True)
+    sc_id = models.TextField(null=True)
+    village_id = models.TextField(null=True)
+
+    class Meta(object):
+        abstract = True
+
+
+class Woman(LocationDenormalizedModel):
+    """Represents a woman registered in the AAA Convergence program.
+
+    This woman may have an eligible_couple or ccs_record case associated with her.
+    ccs_record data will be attached to the CcsRecord model.
+    """
+
+    household_case_id = models.TextField(null=True)
+    person_case_id = models.TextField(primary_key=True)
+
+    opened_on = models.DateField()
+    closed_on = models.DateField(null=True)
+    pregnant_ranges = ArrayField(
+        DateRangeField(),
+        help_text="The ranges in which a ccs_record has been opened and the baby has not been born",
+        null=True
+    )
```

Comment:
This is not a pattern that we've followed before. I have verified that something similar to the following works:

`'2018-02-01'::date <@ any(array[daterange('2015-01-06','2015-02-10'), daterange('2017-10-05','2017-10-11')])`

There is not currently an accessible way to implement that in django, but I believe it should be feasible by writing a custom Django query expression

Reply:
Similar to above, if this is going to be used for filtering the table, have we done any profiling to see how it performs on large tables?

Reply:
I have not yet done profiling, but did check that this can be used with a GiST index. Unfortunately those aren't natively supported in django until 2.0, so I'll need to do something a bit more custom in the meantime